openrouteservice/src/main/java/org/heigit/ors/fastisochrones/partitioning/storage/CellStorage.java

```diff
@@ -0,0 +1,387 @@
+    private IntLongMap cellIdToContourPointerMap;
+    private IntIntMap cellIdToSuperCellMap = new IntIntHashMap();
+    private IntObjectMap<IntHashSet> superCellIdToCellsMap = new IntObjectHashMap<>();
+
+    /**
+     * Instantiates a new Cell storage.
+     *
+     * @param graph                the graph
+     * @param dir                  the dir
+     * @param isochroneNodeStorage the isochrone node storage
+     */
+    public CellStorage(int nodeCount, Directory dir, IsochroneNodeStorage isochroneNodeStorage) {
+        this.isochroneNodeStorage = isochroneNodeStorage;
+        cells = dir.find("cells");
+        byteCount = 4;
+        this.nodeCount = nodeCount;
+    }
+
+    @Override
+    public boolean loadExisting() {
+        if (cells.loadExisting()) {
+            int cellCount = cells.getHeader(0);
+            nodeIndexOffset = cellCount * 12;
+            contourIndexOffset = 2 * cellCount * 18;
+            cellIdToNodesPointerMap = new IntLongHashMap(cellCount);
+            cellIdToContourPointerMap = new IntLongHashMap(cellCount);
+            fillCellIdToNodesPointerMap();
+            fillCellIdToContourPointerMap();
+            if (isSupercellsEnabled()) {
+                fillSuperCellMap();
+                fillCellIdToSuperCellMap();
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Init.
+     */
+    public void init() {
+        cells.create(1000);
+        int cellCount = isochroneNodeStorage.getCellIds().size();
+        cellIdToNodesPointerMap = new IntLongHashMap(cellCount);
+        cellIdToContourPointerMap = new IntLongHashMap(cellCount);
+        cellIdToSuperCellMap = new IntIntHashMap(cellCount);
+    }
+
+    /**
+     * Iterate over all nodes in graph and create a mapping of cells -> nodeIds.
+     * Create a mapping of cellId -> pointer, as the NodeId count per cell is different and the storage irregular.
+     * Store both maps in the storage.
+     */
+    public void calcCellNodesMap() {
+        IntObjectMap<IntHashSet> cellIdToNodesMap = new IntObjectHashMap<>(isochroneNodeStorage.getCellIds().size());
+        //Calc a hashmap of the cells
+        for (int node = 0; node < nodeCount; node++) {
+            int cellId = isochroneNodeStorage.getCellId(node);
+            if (!cellIdToNodesMap.containsKey(cellId))
+                cellIdToNodesMap.put(cellId, new IntHashSet());
+            cellIdToNodesMap.get(cellId).add(node);
+        }
+        int cellCount = cellIdToNodesMap.size();
+        cells.setHeader(0, cellCount);
+        // 2 12-byte pointer sets for each cellId
+        // Store pointers in front that point to where the nodes for each cell start
+        nodeIndexOffset = cellCount * 12;
+        //There are more contours than cells because of supercell contours
+        contourIndexOffset = 2 * cellCount * 18;
+        long nodePointer = (long) contourIndexOffset;
+
+        //Put all the cell nodes in the storage
```

Comment:
Because it means we can garbage collect the hashmap and only have condensed data in the storage. This saves RAM, as the hashmap has to store additional data.

Reply:
Fair enough!